### PR TITLE
CGI script for Tailscale App on Synology

### PR DESCRIPTION
**Pull request by crawshaw**
Lets a user login using the Synology GUI.

### CODE REVIEW

**Comment by crawshaw** [comment]
For #24 and tailscale/tailscale#48.

File: src/config [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+	".url": {
+		"SYNO.SDS.Tailscale": {
+			"type": "url",
+			"version": "1.4.5",
```
Comment (bradfitz) [inline src/config:5]: can you use variables like ${VERSION} here from src/INFO.sh? Probably not? Should we auto-generate this file?

File: src/index.cgi.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
+package main
```
Comment (bradfitz) [inline src/index.cgi.go:1]: copyright header, package doc.

File: src/index.cgi.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+			args[i] = strings.Replace(v, "<", "&lt;", -1)
+		}
+		fmt.Fprintf(buf, format+"\n", args...)
+	}
+	wr("<!doctype html>")
+	wr("<html><title>Tailscale Synology App</title><body>")
+	wr("<h1>Tailscale</h1>")
+	wr(`<div style="float:right;">%s</div>`, user)
+	wr("<table>")
+	wr("<tr><th>Status:</th><td>%s</td></tr>", s.BackendState)
```
Comment (bradfitz) [inline src/index.cgi.go:58]: at least html.EscapeString(...) these

File: src/index.cgi.go [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+	wr("</table>")
+
+	wr(`<form method="POST">`)
+	wr(`<p><input type="submit" value="Log in…"></p>`)
+	wr(`</form>`)
+	url := s.AuthURL
+	if r.FormValue("action") == "login" {
+		_, url, err = tailscaleUp()
+		if err != nil {
+			wr(`<p style="color: red; font-weight: bold;">%v</p>`, err)
```
Comment (bradfitz) [inline src/index.cgi.go:77]: html.EscapeString

File: src/index.cgi.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+		}
+		if url != "" {
+			s.AuthURL = url
+		}
+	}
+	if s.BackendState != "Running" && url != "" {
+		wr(`<p><b>To authorize this device visit: <a href=%q>%s</a></b></p>`, url, url)
```
Comment (bradfitz) [inline src/index.cgi.go:84]: html.EscapeString

File: src/index.cgi.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	}
+
+	wr("</body></html>")
+
+	w.Write(buf.Bytes())
+}
+
+type status struct {
```
Comment (bradfitz) [inline src/index.cgi.go:92]: if we moved this CGI into the main tree, we could just use the right type easily and they wouldn't get out of sync.

Comment (bradfitz) [reply src/index.cgi.go:92]: what about making this tool be part of `cmd/tailscale` when `+build linux`, as an undocumented `tailscale syno-cgi` subcommand. Then we could hard link it into two places as `foo.cgi` and make cmd/tailscale go into CGI mode if its os.Argv[0] is `foo.cgi`.

File: src/index.cgi.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	return s,nil
+}
+
+func tailscaleUp() (state, url string, err error) {
+	cmd := exec.Command("tailscale", "up", "-json")
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return "", "", fmt.Errorf("tailscaleUp: %v: %s", err, out)
+	}
+	var res struct {
```
Comment (bradfitz) [inline src/index.cgi.go:120]: same

File: build-package.sh [hunk 8]
```diff
@@ -28,16 +28,21 @@ make_inner_pkg() {
...
-  cp -a ${tailscale_dir}/tailscale{,d} ${tmp_dir}/bin/
+  mkdir -p "${tmp_dir}/bin"
+  cp -a ${tailscale_dir}/tailscale{,d} "${tmp_dir}/bin/"
 
-  mkdir -p ${tmp_dir}/conf
-  cp -a src/tailscaled_logrotate ${tmp_dir}/conf/logrotate.conf
+  mkdir -p "${tmp_dir}/ui"
+  cp -a src/config "${tmp_dir}/ui/"
+  cp -a src/PACKAGE_ICON_256.PNG "${tmp_dir}/ui/"
+  GOARCH=$ARCH go build -o "${tmp_dir}/ui/index.cgi" src/index.cgi.go
```
Comment (bradfitz) [inline build-package.sh:37]: GOOS=linux too? for those building on macs?

File: src/INFO.sh [hunk 9]
```diff
@@ -32,13 +32,16 @@ arm)
...
 
 cat <<EOF
-package="tailscale"
+package="Tailscale"
 version="${VERSION}"
 arch="${PLATFORMS}"
 description="Connect all your devices using WireGuard, without the hassle."
 displayname="Tailscale"
-maintainer="nirev"
+maintainer="Tailscale"
```
Comment (bradfitz) [inline src/INFO.sh:40]: Tailscale, Inc.?

File: src/index.cgi.go [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+		return
+	}
+
+	s, err := tailscaleStatus()
+	if err != nil {
+		http.Error(w, err.Error(), 500)
+		log.Fatal("tag 2: %v", err.Error())
+		return
+	}
+	if uid := os.Getuid() ; uid != 0 {
```
Comment (bradfitz) [inline src/index.cgi.go:39]: does this matter? I'm afraid in DSM7 they might start sandboxing these and actually running things as non-root. can we just drop this check?

**Review by bradfitz — COMMENTED** [review]
Nice work!

File: src/config (hunk 1 above)
Comment (crawshaw) [reply src/config]: We should generate it, leaving for a later pass after Maisem's code is in.

File: src/config [hunk 11]
```diff
@@ -0,0 +1,13 @@
+{
+	".url": {
+		"SYNO.SDS.Tailscale": {
+			"type": "url",
+			"version": "1.6.0",
```
Comment (maisem) [inline src/config:5]: 1.8.3

**Review by maisem — APPROVED** [review]

**Pull request merged**